Mathlib/Algebra/Lie/Derivation/Basic.lean

```diff
@@ -0,0 +1,311 @@
+  rfl
+
+theorem sub_apply {D1 D2 : LieDerivation R L M} : (D1 - D2) a = D1 a - D2 a :=
+  rfl
+
+section Scalar
+
+/-- A typeclass mixin saying that scalar multiplication and Lie bracket are left commutative. -/
+class SMulBracketCommClass (S L α : Type*) [SMul S α] [LieRing L] [AddCommGroup α]
+    [LieRingModule L α] : Prop where
+  /-- `•` and `⁅⬝, ⬝⁆`  are left commutative -/
+  smul_bracket_comm : ∀ (s : S) (l : L) (a : α), s • ⁅l, a⁆ = ⁅l, s • a⁆
+
+variable {S T : Type*}
+variable [Monoid S] [DistribMulAction S M] [SMulCommClass R S M] [SMulBracketCommClass S L M]
+variable [Monoid T] [DistribMulAction T M] [SMulCommClass R T M] [SMulBracketCommClass T L M]
+
+instance : SMul S (LieDerivation R L M) :=
+  ⟨fun r D =>
+    { toLinearMap := r • D
+      leibniz' := fun a b => by simp only [LinearMap.smul_apply, coeFn_coe, apply_lie_eq_sub,
+        smul_sub, SMulBracketCommClass.smul_bracket_comm] }⟩
+
+@[simp]
+theorem coe_smul (r : S) (D : LieDerivation R L M) : ⇑(r • D) = r • ⇑D :=
+  rfl
+
+@[simp]
+theorem coe_smul_linearMap (r : S) (D : LieDerivation R L M) : ↑(r • D) = r • (D : L →ₗ[R] M) :=
+  rfl
+
+theorem smul_apply (r : S) (D : LieDerivation R L M) : (r • D) a = r • D a :=
+  rfl
+
+instance instSMulNat : SMulBracketCommClass ℕ L M := ⟨fun s l a => (lie_nsmul l a s).symm⟩
+
+instance instSMulInt : SMulBracketCommClass ℤ L M := ⟨fun s l a => (lie_zsmul l a s).symm⟩
+
+instance instAddCommGroup : AddCommGroup (LieDerivation R L M) :=
+  coe_injective.addCommGroup _ coe_zero coe_add coe_neg coe_sub (fun _ _ => rfl) fun _ _ => rfl
+
+/-- `coe_fn` as an `AddMonoidHom`. -/
+def coeFnAddMonoidHom : LieDerivation R L M →+ L → M where
+  toFun := (↑)
+  map_zero' := coe_zero
+  map_add' := coe_add
+
+instance : DistribMulAction S (LieDerivation R L M) :=
+  Function.Injective.distribMulAction coeFnAddMonoidHom coe_injective coe_smul
+
+instance [SMul S T] [IsScalarTower S T M] : IsScalarTower S T (LieDerivation R L M) :=
+  ⟨fun _ _ _ => ext fun _ => smul_assoc _ _ _⟩
+
+instance [SMulCommClass S T M] : SMulCommClass S T (LieDerivation R L M) :=
+  ⟨fun _ _ _ => ext fun _ => smul_comm _ _ _⟩
+
+end Scalar
+
+instance instModule {S : Type*} [Semiring S] [Module S M] [SMulCommClass R S M]
+    [SMulBracketCommClass S L M] : Module S (LieDerivation R L M) :=
+  Function.Injective.module S coeFnAddMonoidHom coe_injective coe_smul
+
+section RestrictScalars
+
+variable {S : Type*} [CommRing S]
+variable [LieAlgebra S L] [Module S M] [LieModule S L M]
+variable [LinearMap.CompatibleSMul L M R S]
+variable (R)
+
+/-- If `L` is both a Lie `R`-algebra and a Lie `S`-algebra, `M` is both an `R`-module and an
+`S`-module, then a Lie `S`-derivation `L → M`, which is `R`-linear is also a Lie `R`-derivation. -/
+protected def restrictScalars (d : LieDerivation S L M) : LieDerivation R L M where
+  leibniz' := d.leibniz'
+  toLinearMap := d.toLinearMap.restrictScalars R
+
+end RestrictScalars
```

Comment:
I suggest dropping this: the day we would use this is very distant (if it ever comes):
```suggestion
```